src/array.jl

```diff
@@ -22,7 +22,7 @@ function unsafe_free!(xs::CuArray)
   return
 end
 
-Base.cconvert(::Type{Ptr{T}}, x::CuArray{T}) where T = x.buf
+Base.cconvert(::Type{<:Ptr}, x::CuArray) = x.buf
```

Comment:
I can see that we'd want to allow `convert(Ptr, xs)`, but would we want things like `convert(Ptr{String}, ::CuArray{Int})` to go through?

I guess as this is not that user-facing it may not matter.

Reply:
I didn't want to, but it currently the `cudnn` wrappers pass a CuArray to `ccall` expecting `Ptr{Void}`:
```julia
function cudnnSoftmaxForward(handle,algo,mode,alpha,xDesc,x,beta,yDesc,y)
    @check ccall((:cudnnSoftmaxForward,libcudnn),cudnnStatus_t,(cudnnHandle_t,cudnnSoftmaxAlgorithm_t,cudnnSoftmaxMode_t,Ptr{Void},cudnnTensorDescriptor_t,Ptr{Void},Ptr{Void},cudnnTensorDescriptor_t,Ptr{Void}),handle,algo,mode,alpha,xDesc,x,beta,yDesc,y)
end

function cudnnSoftmaxForward(src::CuArray, dest::CuArray=src;
                             handle=cudnnHandle,
                             algorithm=CUDNN_SOFTMAX_ACCURATE, # or CUDNN_SOFTMAX_FAST
                             mode=CUDNN_SOFTMAX_MODE_INSTANCE, # or CUDNN_SOFTMAX_MODE_CHANNEL
                             alpha=1.0, beta=0.0)
    cudnnSoftmaxForward(handle, algorithm, mode,
                        cptr(alpha, src), TD(src,4), src,
                        cptr(beta, dest), TD(dest,4), dest)
    return dest
end
```

Either we type the inner `cudnnSoftmaxForward`, or manually convert to a typed pointer in the outer one, neither seems nice to me. Other suggestions?

Reply:
Will plain `Ptr` work in the ccall? If not, perhaps we can just make `convert(Ptr{Void}, xs)` a special case that ignores eltype, which aligns well with the meaning in C anyway.

Reply:
Plain `Ptr` doesn't work with `unsafe_convert`, but the special case rule is a better solution indeed.